--- /Users/yoyofr/Documents/Dev/rewamp/packages/rewamp_audio/third_party/libkss/upstream/src/kss/bgm2kss.c	2026-07-06 09:48:52
+++ src/kss/bgm2kss.c	2026-07-06 09:44:22
@@ -14,7 +14,7 @@
     0xCD, 0x20, 0x60,       /* CALL	6020H */
     0x3E, 0x3F,             /* LD	A,3FH */
     0x32, 0x10, 0x60,       /* LD	(6010H),A	; OPLL SLOT */
-    0x32, 0x11, 0x60,       /* LD	(6011H),A	; SCC SLOT */
+    0x32, 0x11, 0x60,       /* LD	(6011H),A	; SCCKSS SLOT */
     0x21, 0x07, 0x80,       /* LD HL,A007H ; DATA ADDRESS */
     0xED, 0x5B, 0x01, 0x80, /* LD	DE,(A001H) ; COMPILE ADDRESS */
     /* INIT FLAG WORK */
