--- /Users/yoyofr/Documents/Dev/rewamp/packages/rewamp_audio/third_party/libkss/upstream/src/filters/dc_filter.h	2026-07-06 09:48:51
+++ src/filters/dc_filter.h	2026-07-06 09:44:22
@@ -10,10 +10,12 @@
 typedef struct _tagDCF {
   uint32_t enable;
   double weight;
+    //YOYOFR
   double in, out;
 } DCF;
 
 DCF *DCF_new(void);
+//YOYOFR
 int32_t DCF_calc(DCF *, int32_t data);
 void DCF_delete(DCF *);
 void DCF_reset(DCF *, double rate);
