--- third_party/libkss/upstream/modules/emu76489/emu76489.c	2026-07-06 09:48:52
+++ third_party/libkss/modules/emu76489/emu76489.c	2026-07-06 13:34:41
@@ -20,6 +20,11 @@
 #include <string.h>
 #include "emu76489.h"
 
+//TODO:  MODIZER changes start / YOYOFR
+#include "../../../../src/ModizerVoicesData.h"
+//TODO:  MODIZER changes end / YOYOFR
+
+
 static uint32_t voltbl[16] = {
   0xff, 0xcb, 0xa1, 0x80, 0x65, 0x50, 0x40, 0x33, 0x28, 0x20, 0x19, 0x14, 0x10, 0x0c, 0x0a, 0x00
 };
@@ -227,7 +232,32 @@
 static inline int16_t
 mix_output (SNG * sng) 
 {
-  sng->out = sng->ch_out[0] + sng->ch_out[1] + sng->ch_out[2] + sng->ch_out[3];
+    //TODO:  MODIZER changes start / YOYOFR
+    sng->out=0;
+    for (int i=0;i<4;i++)
+        if (!(generic_mute_mask&((int64_t)1<<(m_voicesForceOfs+i)))) sng->out += sng->ch_out[i];
+//  sng->out = sng->ch_out[0] + sng->ch_out[1] + sng->ch_out[2] + sng->ch_out[3];
+  
+    
+    int64_t smplIncr=(int64_t)(1<<MODIZER_OSCILLO_OFFSET_FIXEDPOINT)/m_voice_current_rateratio;
+    if (m_voicesForceOfs>=0) {
+        int val=0;
+        for (int i = 0; i < 4; i++) {
+            if (!(generic_mute_mask&((int64_t)1<<(m_voicesForceOfs+i)))) val=sng->ch_out[i];
+            else val=0;
+            int64_t ofs_start=m_voice_current_ptr[m_voicesForceOfs+i];
+            int64_t ofs_end=ofs_start+smplIncr;
+            for (;;) {
+                m_voice_buff[m_voicesForceOfs+i][(ofs_start>>MODIZER_OSCILLO_OFFSET_FIXEDPOINT)&(SOUND_BUFFER_SIZE_SAMPLE*4*2-1)]=LIMIT8((val>>5));
+                ofs_start+=1<<MODIZER_OSCILLO_OFFSET_FIXEDPOINT;
+                if (ofs_start>=ofs_end) break;
+            }
+            while ((ofs_end>>MODIZER_OSCILLO_OFFSET_FIXEDPOINT)>=SOUND_BUFFER_SIZE_SAMPLE*4*2) ofs_end-=(SOUND_BUFFER_SIZE_SAMPLE*4*2<<MODIZER_OSCILLO_OFFSET_FIXEDPOINT);
+            m_voice_current_ptr[m_voicesForceOfs+i]=ofs_end;
+        }
+    }
+    //TODO:  MODIZER changes end / YOYOFR
+    
   return (int16_t) sng->out;
 }
 
@@ -258,22 +288,41 @@
 
   out[0] = out[1] = 0;
   if((sng->stereo>>4)&0x08) {
-    out[0] += sng->ch_out[3];
+      if (!(generic_mute_mask&((int64_t)1<<(m_voicesForceOfs+3)))) out[0] += sng->ch_out[3];
   }
   if(sng->stereo & 0x08) {
-    out[1] += sng->ch_out[3];
+      if (!(generic_mute_mask&((int64_t)1<<(m_voicesForceOfs+3)))) out[1] += sng->ch_out[3];
   }
-
+        
   for (i = 0; i < 3; i++)
   {
     if ((sng->stereo >> (i+4)) & 0x01) {
-      out[0] += sng->ch_out[i];
+        if (!(generic_mute_mask&((int64_t)1<<(m_voicesForceOfs+i)))) out[0] += sng->ch_out[i];
     }
     if ((sng->stereo >> i) & 0x01) {
-      out[1] += sng->ch_out[i];
+        if (!(generic_mute_mask&((int64_t)1<<(m_voicesForceOfs+i)))) out[1] += sng->ch_out[i];
     }
   }
-
+    
+    //TODO:  MODIZER changes start / YOYOFR
+    int64_t smplIncr=(int64_t)(1<<MODIZER_OSCILLO_OFFSET_FIXEDPOINT)/m_voice_current_rateratio;
+    if (m_voicesForceOfs>=0) {
+        int val=0;
+        for (int i = 0; i < 4; i++) {
+            if (!(generic_mute_mask&((int64_t)1<<(m_voicesForceOfs+i)))) val=sng->ch_out[i];
+            else val=0;
+            int64_t ofs_start=m_voice_current_ptr[m_voicesForceOfs+i];
+            int64_t ofs_end=ofs_start+smplIncr;
+            for (;;) {
+                m_voice_buff[m_voicesForceOfs+i][(ofs_start>>MODIZER_OSCILLO_OFFSET_FIXEDPOINT)&(SOUND_BUFFER_SIZE_SAMPLE*4*2-1)]=LIMIT8((val>>5));
+                ofs_start+=1<<MODIZER_OSCILLO_OFFSET_FIXEDPOINT;
+                if (ofs_start>=ofs_end) break;
+            }
+            while ((ofs_end>>MODIZER_OSCILLO_OFFSET_FIXEDPOINT)>=SOUND_BUFFER_SIZE_SAMPLE*4*2) ofs_end-=(SOUND_BUFFER_SIZE_SAMPLE*4*2<<MODIZER_OSCILLO_OFFSET_FIXEDPOINT);
+            m_voice_current_ptr[m_voicesForceOfs+i]=ofs_end;
+        }
+    }
+    //TODO:  MODIZER changes end / YOYOFR
 }
 
 void
