--- /Users/yoyofr/Documents/Dev/rewamp/packages/rewamp_audio/third_party/libkss/upstream/src/kss/kss.c	2026-07-06 09:48:52
+++ src/kss/kss.c	2026-07-06 09:44:22
@@ -51,7 +51,7 @@
     return MPK106DATA;
   else if (KSS_isMPK103data(data, size))
     return MPK103DATA;
-  else if (!strncmp("KSCC", (const char *)data, 4))
+  else if (!strncmp("KSCCKSS", (const char *)data, 4))
     return KSSDATA;
   else if (!strncmp("KSSX", (const char *)data, 4))
     return KSSDATA;
@@ -141,7 +141,7 @@
 static void scan_info(KSS *kss) {
   uint32_t i, j, k;
 
-  if (strncmp("KSCC", (const char *)(kss->data), 4) == 0) {
+  if (strncmp("KSCCKSS", (const char *)(kss->data), 4) == 0) {
     kss->kssx = 0;
     get_legacy_header(kss);
     check_device(kss, kss->device_flag);
