--- /Users/yoyofr/Documents/Dev/rewamp/packages/rewamp_audio/third_party/libkss/upstream/modules/emu2413/kss_emu2413.h	2026-07-06 09:48:52
+++ modules/emu2413/kss_emu2413.h	2026-07-06 09:44:22
@@ -1,5 +1,5 @@
-#ifndef _EMU2413_H_
-#define _EMU2413_H_
+#ifndef _EMU2413KSS_H_
+#define _EMU2413KSS_H_
 
 #include <stdint.h>
 
@@ -7,17 +7,17 @@
 extern "C" {
 #endif
 
-#define OPLL_DEBUG 0
+#define OPLLKSS_DEBUG 0
 
-enum OPLL_TONE_ENUM { OPLL_2413_TONE = 0, OPLL_VRC7_TONE = 1, OPLL_281B_TONE = 2 };
+enum OPLLKSS_TONE_ENUM { OPLLKSS_2413_TONE = 0, OPLLKSS_VRC7_TONE = 1, OPLLKSS_281B_TONE = 2 };
 
 /* voice data */
-typedef struct __OPLL_PATCH {
+typedef struct __OPLLKSS_PATCH {
   uint32_t TL, FB, EG, ML, AR, DR, SL, RR, KR, KL, AM, PM, WS;
-} OPLL_PATCH;
+} OPLLKSS_PATCH;
 
 /* slot */
-typedef struct __OPLL_SLOT {
+typedef struct __OPLLKSS_SLOT {
   uint8_t number;
 
   /* type flags:
@@ -27,7 +27,7 @@
    */
   uint8_t type;
 
-  OPLL_PATCH *patch; /* voice parameter */
+  OPLLKSS_PATCH *patch; /* voice parameter */
 
   /* slot output */
   int32_t output[2]; /* output value, latest and previous. */
@@ -55,36 +55,36 @@
 
   uint32_t update_requests; /* flags to debounce update */
 
-#if OPLL_DEBUG
+#if OPLLKSS_DEBUG
   uint8_t last_eg_state;
 #endif
-} OPLL_SLOT;
+} OPLLKSS_SLOT;
 
 /* mask */
-#define OPLL_MASK_CH(x) (1 << (x))
-#define OPLL_MASK_HH (1 << (9))
-#define OPLL_MASK_CYM (1 << (10))
-#define OPLL_MASK_TOM (1 << (11))
-#define OPLL_MASK_SD (1 << (12))
-#define OPLL_MASK_BD (1 << (13))
-#define OPLL_MASK_RHYTHM (OPLL_MASK_HH | OPLL_MASK_CYM | OPLL_MASK_TOM | OPLL_MASK_SD | OPLL_MASK_BD)
+#define OPLLKSS_MASK_CH(x) (1 << (x))
+#define OPLLKSS_MASK_HH (1 << (9))
+#define OPLLKSS_MASK_CYM (1 << (10))
+#define OPLLKSS_MASK_TOM (1 << (11))
+#define OPLLKSS_MASK_SD (1 << (12))
+#define OPLLKSS_MASK_BD (1 << (13))
+#define OPLLKSS_MASK_RHYTHM (OPLLKSS_MASK_HH | OPLLKSS_MASK_CYM | OPLLKSS_MASK_TOM | OPLLKSS_MASK_SD | OPLLKSS_MASK_BD)
 
 /* rate conveter */
-typedef struct __OPLL_RateConv {
+typedef struct __OPLLKSS_RateConv {
   int ch;
   double timer;
   double f_ratio;
   int16_t *sinc_table;
   int16_t **buf;
-} OPLL_RateConv;
+} OPLLKSS_RateConv;
 
-OPLL_RateConv *OPLL_RateConv_new(double f_inp, double f_out, int ch);
-void OPLL_RateConv_reset(OPLL_RateConv *conv);
-void OPLL_RateConv_putData(OPLL_RateConv *conv, int ch, int16_t data);
-int16_t OPLL_RateConv_getData(OPLL_RateConv *conv, int ch);
-void OPLL_RateConv_delete(OPLL_RateConv *conv);
+OPLLKSS_RateConv *OPLLKSS_RateConv_new(double f_inp, double f_out, int ch);
+void OPLLKSS_RateConv_reset(OPLLKSS_RateConv *conv);
+void OPLLKSS_RateConv_putData(OPLLKSS_RateConv *conv, int ch, int16_t data);
+int16_t OPLLKSS_RateConv_getData(OPLLKSS_RateConv *conv, int ch);
+void OPLLKSS_RateConv_delete(OPLLKSS_RateConv *conv);
 
-typedef struct __OPLL {
+typedef struct __OPLLKSS {
   uint32_t clk;
   uint32_t rate;
 
@@ -112,8 +112,8 @@
   uint8_t short_noise;
 
   int32_t patch_number[9];
-  OPLL_SLOT slot[18];
-  OPLL_PATCH patch[19 * 2];
+  OPLLKSS_SLOT slot[18];
+  OPLLKSS_PATCH patch[19 * 2];
 
   uint8_t pan[16];
   float pan_fine[16][2];
@@ -126,27 +126,27 @@
 
   int16_t mix_out[2];
 
-  OPLL_RateConv *conv;
-} OPLL;
+  OPLLKSS_RateConv *conv;
+} OPLLKSS;
 
-OPLL *OPLL_new(uint32_t clk, uint32_t rate);
-void OPLL_delete(OPLL *);
+OPLLKSS *OPLLKSS_new(uint32_t clk, uint32_t rate);
+void OPLLKSS_delete(OPLLKSS *);
 
-void OPLL_reset(OPLL *);
-void OPLL_resetPatch(OPLL *, uint8_t);
+void OPLLKSS_reset(OPLLKSS *);
+void OPLLKSS_resetPatch(OPLLKSS *, uint8_t);
 
 /**
  * Set output wave sampling rate.
  * @param rate sampling rate. If clock / 72 (typically 49716 or 49715 at 3.58MHz) is set, the internal rate converter is
  * disabled.
  */
-void OPLL_setRate(OPLL *opll, uint32_t rate);
+void OPLLKSS_setRate(OPLLKSS *opll, uint32_t rate);
 
 /**
  * Set internal calcuration quality. Currently no effects, just for compatibility.
  * >= v1.0.0 always synthesizes internal output at clock/72 Hz.
  */
-void OPLL_setQuality(OPLL *opll, uint8_t q);
+void OPLLKSS_setQuality(OPLLKSS *opll, uint8_t q);
 
 /**
  * Set pan pot (extra function - not YM2413 chip feature)
@@ -158,7 +158,7 @@
  *            +-- bit 0: enable Right output
  * ```
  */
-void OPLL_setPan(OPLL *opll, uint32_t ch, uint8_t pan);
+void OPLLKSS_setPan(OPLLKSS *opll, uint32_t ch, uint8_t pan);
 
 /**
  * Set fine-grained panning
@@ -166,69 +166,69 @@
  * @param pan output strength of left/right channel.
  *            pan[0]: left, pan[1]: right. pan[0]=pan[1]=1.0f for center.
  */
-void OPLL_setPanFine(OPLL *opll, uint32_t ch, float pan[2]);
+void OPLLKSS_setPanFine(OPLLKSS *opll, uint32_t ch, float pan[2]);
 
 /**
  * Set chip type. If vrc7 is selected, r#14 is ignored.
  * This method not change the current ROM patch set.
- * To change ROM patch set, use OPLL_resetPatch.
+ * To change ROM patch set, use OPLLKSS_resetPatch.
  * @param type 0:YM2413 1:VRC7
  */
-void OPLL_setChipType(OPLL *opll, uint8_t type);
+void OPLLKSS_setChipType(OPLLKSS *opll, uint8_t type);
 
-void OPLL_writeIO(OPLL *opll, uint32_t reg, uint8_t val);
-void OPLL_writeReg(OPLL *opll, uint32_t reg, uint8_t val);
+void OPLLKSS_writeIO(OPLLKSS *opll, uint32_t reg, uint8_t val);
+void OPLLKSS_writeReg(OPLLKSS *opll, uint32_t reg, uint8_t val);
 
 /**
  * Calculate one sample
  */
-int16_t OPLL_calc(OPLL *opll);
+int16_t OPLLKSS_calc(OPLLKSS *opll);
 
 /**
  * Calulate stereo sample
  */
-void OPLL_calcStereo(OPLL *opll, int32_t out[2]);
+void OPLLKSS_calcStereo(OPLLKSS *opll, int32_t out[2]);
 
-void OPLL_setPatch(OPLL *, const uint8_t *dump);
-void OPLL_copyPatch(OPLL *, int32_t, OPLL_PATCH *);
+void OPLLKSS_setPatch(OPLLKSS *, const uint8_t *dump);
+void OPLLKSS_copyPatch(OPLLKSS *, int32_t, OPLLKSS_PATCH *);
 
 /**
  * Force to refresh.
  * External program should call this function after updating patch parameters.
  */
-void OPLL_forceRefresh(OPLL *);
+void OPLLKSS_forceRefresh(OPLLKSS *);
 
-void OPLL_dumpToPatch(const uint8_t *dump, OPLL_PATCH *patch);
-void OPLL_patchToDump(const OPLL_PATCH *patch, uint8_t *dump);
-void OPLL_getDefaultPatch(int32_t type, int32_t num, OPLL_PATCH *);
+void OPLLKSS_dumpToPatch(const uint8_t *dump, OPLLKSS_PATCH *patch);
+void OPLLKSS_patchToDump(const OPLLKSS_PATCH *patch, uint8_t *dump);
+void OPLLKSS_getDefaultPatch(int32_t type, int32_t num, OPLLKSS_PATCH *);
 
 /**
  *  Set channel mask
- *  @param mask mask flag: OPLL_MASK_* can be used.
- *  - bit 0..8: mask for ch 1 to 9 (OPLL_MASK_CH(i))
- *  - bit 9: mask for Hi-Hat (OPLL_MASK_HH)
- *  - bit 10: mask for Top-Cym (OPLL_MASK_CYM)
- *  - bit 11: mask for Tom (OPLL_MASK_TOM)
- *  - bit 12: mask for Snare Drum (OPLL_MASK_SD)
- *  - bit 13: mask for Bass Drum (OPLL_MASK_BD)
+ *  @param mask mask flag: OPLLKSS_MASK_* can be used.
+ *  - bit 0..8: mask for ch 1 to 9 (OPLLKSS_MASK_CH(i))
+ *  - bit 9: mask for Hi-Hat (OPLLKSS_MASK_HH)
+ *  - bit 10: mask for Top-Cym (OPLLKSS_MASK_CYM)
+ *  - bit 11: mask for Tom (OPLLKSS_MASK_TOM)
+ *  - bit 12: mask for Snare Drum (OPLLKSS_MASK_SD)
+ *  - bit 13: mask for Bass Drum (OPLLKSS_MASK_BD)
  */
-uint32_t OPLL_setMask(OPLL *, uint32_t mask);
+uint32_t OPLLKSS_setMask(OPLLKSS *, uint32_t mask);
 
 /**
  * Toggler channel mask flag
  */
-uint32_t OPLL_toggleMask(OPLL *, uint32_t mask);
+uint32_t OPLLKSS_toggleMask(OPLLKSS *, uint32_t mask);
 
 /* for compatibility */
-#define OPLL_set_rate OPLL_setRate
-#define OPLL_set_quality OPLL_setQuality
-#define OPLL_set_pan OPLL_setPan
-#define OPLL_set_pan_fine OPLL_setPanFine
-#define OPLL_calc_stereo OPLL_calcStereo
-#define OPLL_reset_patch OPLL_resetPatch
-#define OPLL_dump2patch OPLL_dumpToPatch
-#define OPLL_patch2dump OPLL_patchToDump
-#define OPLL_setChipMode OPLL_setChipType
+#define OPLLKSS_set_rate OPLLKSS_setRate
+#define OPLLKSS_set_quality OPLLKSS_setQuality
+#define OPLLKSS_set_pan OPLLKSS_setPan
+#define OPLLKSS_set_pan_fine OPLLKSS_setPanFine
+#define OPLLKSS_calc_stereo OPLLKSS_calcStereo
+#define OPLLKSS_reset_patch OPLLKSS_resetPatch
+#define OPLLKSS_dump2patch OPLLKSS_dumpToPatch
+#define OPLLKSS_patch2dump OPLLKSS_patchToDump
+#define OPLLKSS_setChipMode OPLLKSS_setChipType
 
 #ifdef __cplusplus
 }
