--- /Users/yoyofr/Documents/Dev/rewamp/packages/rewamp_audio/third_party/libkss/upstream/src/vm/vm.c	2026-07-06 09:48:51
+++ src/vm/vm.c	2026-07-06 09:44:22
@@ -23,13 +23,13 @@
   }
 
   if (!vm->scc_disable) {
-    if (vm->scc_type == VM_SCC_AUTO && (a == 0x9000 || a == 0xB000 || a == 0xBFFE)) {
+    if (vm->scc_type == VM_SCCKSS_AUTO && (a == 0x9000 || a == 0xB000 || a == 0xBFFE)) {
       if (a == 0x9000 && d != 0x00)
-        SCC_write(vm->scc, a, 0x3f);
+        SCCKSS_write(vm->scc, a, 0x3f);
       else
-        SCC_write(vm->scc, a, d);
+        SCCKSS_write(vm->scc, a, d);
     } else if ((0x9800 <= a && a <= 0x98FF) || (0xB800 <= a && a <= 0xB8FF)) {
-      SCC_write(vm->scc, a, d);
+      SCCKSS_write(vm->scc, a, d);
       LPDETECT_write(vm->lpde, a, d);
     }
   }
@@ -52,11 +52,11 @@
   a &= 0xff;
 
   if ((vm->psg) && ((a & 0xff) == 0xA2))
-    return PSG_readIO(vm->psg);
+    return PSGKSS_readIO(vm->psg);
   else if (a == 0xC1)
-    return OPL_readIO(vm->opl);
+    return OPLKSS_readIO(vm->opl);
   else if (a == 0xC0)
-    return OPL_status(vm->opl);
+    return OPLKSS_status(vm->opl);
 
   return 0xff;
 }
@@ -89,7 +89,7 @@
   case 0xA0:
   case 0xA1:
     if (vm->psg) {
-      PSG_writeIO(vm->psg, a, d);
+      PSGKSS_writeIO(vm->psg, a, d);
       LPDETECT_write(vm->lpde, a, d);
     }
     break;
@@ -119,7 +119,7 @@
   case 0xF0:
   case 0xF1:
     if (vm->opll) {
-      OPLL_writeIO(vm->opll, a, d);
+      OPLLKSS_writeIO(vm->opll, a, d);
       LPDETECT_write(vm->lpde, a, d);
     }
     break;
@@ -127,7 +127,7 @@
   case 0xC0:
   case 0xC1:
     if (vm->opl) {
-      OPL_writeIO(vm->opl, a, d);
+        OPLKSS_writeIO(vm->opl, a, d);
       LPDETECT_write(vm->lpde, a, d);
     }
     break;
@@ -179,18 +179,18 @@
   memset(vm, 0, sizeof(VM));
 
   vm->sng = SNG_new(MSX_CLK, rate);
-  vm->psg = PSG_new(MSX_CLK, rate);
-  PSG_setClockDivider(vm->psg, 1);
-  vm->scc = SCC_new(MSX_CLK, rate);
-  vm->opll = OPLL_new(MSX_CLK, rate);
-  vm->opl = OPL_new(MSX_CLK, rate);
+  vm->psg = PSGKSS_new(MSX_CLK, rate);
+  PSGKSS_setClockDivider(vm->psg, 1);
+  vm->scc = SCCKSS_new(MSX_CLK, rate);
+  vm->opll = OPLLKSS_new(MSX_CLK, rate);
+  vm->opl = OPLKSS_new(MSX_CLK, rate);
   vm->mmap = MMAP_new();
   vm->lpde = LPDETECT_new();
   vm->ram_mode = 0;
   vm->DA8_enable = 0;
   vm->bank_mode = BANK_16K;
-  vm->psg_type = VM_PSG_AUTO;
-  vm->scc_type = VM_SCC_AUTO;
+  vm->psg_type = VM_PSGKSS_AUTO;
+  vm->scc_type = VM_SCCKSS_AUTO;
   vm->opll_type = VM_OPLL_2413;
   assert(vm->mmap);
 
@@ -202,10 +202,10 @@
   MMAP_delete(vm->mmap);
   LPDETECT_delete(vm->lpde);
   SNG_delete(vm->sng);
-  PSG_delete(vm->psg);
-  SCC_delete(vm->scc);
-  OPLL_delete(vm->opll);
-  OPL_delete(vm->opl);
+  PSGKSS_delete(vm->psg);
+  SCCKSS_delete(vm->scc);
+    OPLLKSS_delete(vm->opll);
+    OPLKSS_delete(vm->opl);
   kmevent_free(&vm->kme, vm->vsync_id);
   free(vm);
 }
@@ -223,44 +223,44 @@
 }
 
 void VM_reset_device(VM *vm) {
-  PSG_reset(vm->psg);
-  VM_set_PSG_type(vm, vm->psg_type);
-  SCC_reset(vm->scc);
-  VM_set_SCC_type(vm, vm->scc_type);
-  OPLL_reset(vm->opll);
+  PSGKSS_reset(vm->psg);
+  VM_set_PSGKSS_type(vm, vm->psg_type);
+  SCCKSS_reset(vm->scc);
+  VM_set_SCCKSS_type(vm, vm->scc_type);
+    OPLLKSS_reset(vm->opll);
   VM_set_OPLL_type(vm, vm->opll_type);
-  OPL_reset(vm->opl);
+    OPLKSS_reset(vm->opl);
   VM_set_OPL_type(vm, vm->opl_type);
   SNG_reset(vm->sng);
 }
 
-void VM_set_PSG_type(VM *vm, uint32_t psg_type) {
+void VM_set_PSGKSS_type(VM *vm, uint32_t psg_type) {
   if (!vm->psg)
     return;
 
-  if (psg_type == VM_PSG_AUTO)
-    PSG_setVolumeMode(vm->psg, 0);
-  else if (psg_type == VM_PSG_AY)
-    PSG_setVolumeMode(vm->psg, 2);
-  else if (psg_type == VM_PSG_YM)
-    PSG_setVolumeMode(vm->psg, 1);
+  if (psg_type == VM_PSGKSS_AUTO)
+    PSGKSS_setVolumeMode(vm->psg, 0);
+  else if (psg_type == VM_PSGKSS_AY)
+    PSGKSS_setVolumeMode(vm->psg, 2);
+  else if (psg_type == VM_PSGKSS_YM)
+    PSGKSS_setVolumeMode(vm->psg, 1);
   vm->psg_type = psg_type;
 }
 
-void VM_set_SCC_type(VM *vm, uint32_t scc_type) {
+void VM_set_SCCKSS_type(VM *vm, uint32_t scc_type) {
   if (!vm->scc)
     return;
 
-  if (scc_type == VM_SCC_AUTO) {
-    SCC_set_type(vm->scc, SCC_ENHANCED);
-    SCC_write(vm->scc, 0x9000, 0x3f);
-  } else if (scc_type == VM_SCC_STANDARD) {
-    SCC_set_type(vm->scc, SCC_STANDARD);
-    SCC_write(vm->scc, 0x9000, 0x3f);
-  } else if (scc_type == VM_SCC_ENHANCED) {
-    SCC_set_type(vm->scc, SCC_ENHANCED);
-    SCC_write(vm->scc, 0xBFFE, 0x20);
-    SCC_write(vm->scc, 0xB000, 0x80);
+  if (scc_type == VM_SCCKSS_AUTO) {
+    SCCKSS_set_type(vm->scc, SCCKSS_ENHANCED);
+    SCCKSS_write(vm->scc, 0x9000, 0x3f);
+  } else if (scc_type == VM_SCCKSS_STANDARD) {
+    SCCKSS_set_type(vm->scc, SCCKSS_STANDARD);
+    SCCKSS_write(vm->scc, 0x9000, 0x3f);
+  } else if (scc_type == VM_SCCKSS_ENHANCED) {
+    SCCKSS_set_type(vm->scc, SCCKSS_ENHANCED);
+    SCCKSS_write(vm->scc, 0xBFFE, 0x20);
+    SCCKSS_write(vm->scc, 0xB000, 0x80);
   }
   vm->scc_type = scc_type;
 }
@@ -270,11 +270,11 @@
     return;
 
   if (opll_type == VM_OPLL_2413)
-    OPLL_reset_patch(vm->opll, OPLL_2413_TONE);
+      OPLLKSS_reset_patch(vm->opll,OPLLKSS_2413_TONE);
   else if (opll_type == VM_OPLL_VRC7)
-    OPLL_reset_patch(vm->opll, OPLL_VRC7_TONE);
+      OPLLKSS_reset_patch(vm->opll,OPLLKSS_VRC7_TONE);
   else if (opll_type == VM_OPLL_281B)
-    OPLL_reset_patch(vm->opll, OPLL_281B_TONE);
+      OPLLKSS_reset_patch(vm->opll,OPLLKSS_281B_TONE);
   else
     ;
   vm->opll_type = opll_type;
@@ -386,4 +386,4 @@
 
 void VM_write_io(VM *vm, uint32_t a, uint32_t d) {
   iowrite(vm, a, d);
-}
\ No newline at end of file
+}
