diff --git a/emu/cores/ym3438.c b/emu/cores/ym3438.c
index 719b77b..75c4bb3 100644
--- a/emu/cores/ym3438.c
+++ b/emu/cores/ym3438.c
@@ -37,6 +37,14 @@
 #include "ym3438.h"
 #include "ym3438_int.h"
 
+//TODO:  MODIZER changes start / YOYOFR
+#include "../../../../../src/ModizerVoicesData.h"
+static int m_voice_ofs=-1;
+static int64_t smplIncr;
+static int cur_chan=0;
+//TODO:  MODIZER changes end / YOYOFR
+
+
 // superctr's MegaDrive model 1 filter
 #define FILTER_CUTOFF 0.512331301282628 // 5894Hz  single pole IIR low pass
 #define FILTER_CUTOFF_I (1-FILTER_CUTOFF)
@@ -1348,7 +1356,7 @@ void NOPN2_Clock(ym3438_t *chip, Bit32s *buffer)
         chip->pg_block = chip->block[(chip->channel + 1) % 6];
         chip->pg_kcode = chip->kcode[(chip->channel + 1) % 6];
     }
-
+    
     NOPN2_UpdateLFO(chip);
     NOPN2_DoRegWrite(chip);
     chip->cycles = (chip->cycles + 1) % 24;
@@ -1356,6 +1364,9 @@ void NOPN2_Clock(ym3438_t *chip, Bit32s *buffer)
 
     buffer[0] = chip->mol;
     buffer[1] = chip->mor;
+    
+    chip->chan_out[cur_chan]+=chip->mol+chip->mor; //YOYOFR
+    
 
     if (chip->status_time)
         chip->status_time--;
@@ -1497,6 +1508,9 @@ void NOPN2_GenerateResampled(ym3438_t *chip, Bit32s *buf)
     
     while (chip->samplecnt >= chip->rateratio)
     {
+        //YOYOFR
+        for (int jj=0;jj<6;jj++) chip->chan_out[jj]=0;
+        
         chip->oldsamples[0] = chip->samples[0];
         chip->oldsamples[1] = chip->samples[1];
         chip->samples[0] = chip->samples[1] = 0;
@@ -1505,21 +1519,27 @@ void NOPN2_GenerateResampled(ym3438_t *chip, Bit32s *buf)
             switch (chip->cycles >> 2)
             {
             case 0: // Ch 2
+                    cur_chan=1;//YOYOFR
                 mute = chip->mute[1];
                 break;
             case 1: // Ch 6, DAC
+                    cur_chan=5;//YOYOFR
                 mute = chip->mute[5 + chip->dacen];
                 break;
             case 2: // Ch 4
+                    cur_chan=3;//YOYOFR
                 mute = chip->mute[3];
                 break;
             case 3: // Ch 1
+                    cur_chan=0;//YOYOFR
                 mute = chip->mute[0];
                 break;
             case 4: // Ch 5
+                    cur_chan=4;//YOYOFR
                 mute = chip->mute[4];
                 break;
             case 5: // Ch 3
+                    cur_chan=2;//YOYOFR
                 mute = chip->mute[2];
                 break;
             default:
@@ -1532,7 +1552,7 @@ void NOPN2_GenerateResampled(ym3438_t *chip, Bit32s *buf)
                 chip->samples[0] += buffer[0];
                 chip->samples[1] += buffer[1];
             }
-
+            
             while (chip->writebuf[chip->writebuf_cur].time <= chip->writebuf_samplecnt)
             {
                 if (!(chip->writebuf[chip->writebuf_cur].port & 0x04))
@@ -1546,6 +1566,29 @@ void NOPN2_GenerateResampled(ym3438_t *chip, Bit32s *buf)
             }
             chip->writebuf_samplecnt++;
         }
+        
+        
+        //TODO:  MODIZER changes start / YOYOFR
+        if (m_voice_ofs>=0) {
+            for (int jj=0;jj<6;jj++) {
+                int64_t ofs_start=m_voice_current_ptr[m_voice_ofs+jj];
+                int64_t ofs_end=(m_voice_current_ptr[m_voice_ofs+jj]+smplIncr);
+                
+                
+                if (!chip->mute[jj] && (ofs_end>ofs_start))
+                for (;;) {
+                    m_voice_buff[m_voice_ofs+jj][(ofs_start>>MODIZER_OSCILLO_OFFSET_FIXEDPOINT)&(SOUND_BUFFER_SIZE_SAMPLE*4*2-1)]=LIMIT8(((chip->chan_out[jj])>>4));
+                    
+                    ofs_start+=1<<MODIZER_OSCILLO_OFFSET_FIXEDPOINT;
+                    if (ofs_start>=ofs_end) break;
+                }
+                
+                m_voice_current_ptr[m_voice_ofs+jj]=ofs_end;
+            }
+        }
+        //YOYOFR
+        
+        
         if(!chip->use_filter)
         {
             chip->samples[0] *= 11;
@@ -1573,6 +1616,63 @@ void nukedopn2_update(void *chip, UINT32 numsamples, DEV_SMPL **sndptr)
     Bit32s buffer[2];
     smpl = sndptr[0];
     smpr = sndptr[1];
+    
+    //TODO:  MODIZER changes start / YOYOFR
+    //search first voice linked to current chip
+    static INT32 old_out_fm[6]; //YOYOFR
+    static UINT8 old_dacen; //YOYOFR
+    
+    int m_total_channels=6;
+    for (int ii=0;ii<=SOUND_MAXVOICES_BUFFER_FX-m_total_channels;ii++) {
+        if (m_voice_ChipID[ii]==m_voice_current_system) {
+            m_voice_ofs=ii+(m_voice_current_systemSub?m_voice_current_systemPairedOfs:0);
+            m_voice_current_total=m_total_channels;
+            break;
+        }
+    }
+    if (!m_voice_current_samplerate) {
+        m_voice_current_samplerate=44100;
+        //printf("voice sample rate null\n");
+    }
+    smplIncr=(int64_t)44100*(1<<MODIZER_OSCILLO_OFFSET_FIXEDPOINT)/m_voice_current_samplerate;
+    //TODO:  MODIZER changes end / YOYOFR
+
+    
+    //YOYOFR
+    if (m_voice_ofs>=0)
+        for (int ii=0;ii<6;ii++) {
+            if (!(opn2->mute[ii])) {
+                if (opn2->fnum[ii]==0) {
+                    if ((opn2->ch_out[ii]!=old_out_fm[ii])) {
+                        vgm_last_note[ii+m_voice_ofs]=220.0f; //arbitrary choosing A-3
+                        vgm_last_instr[ii+m_voice_ofs]=ii+m_voice_ofs;
+                        if ((ii==5)&&(opn2->dacen)) {
+                            if (!old_dacen) {
+                                vgm_last_vol[ii+m_voice_ofs]=2;
+                            } else vgm_last_vol[ii+m_voice_ofs]=1;
+                        } else {
+                            int newvol=1;//cch[ii]->keyon_triggered+1;
+                            vgm_last_vol[ii+m_voice_ofs]=newvol;
+                        }
+                    }
+                } else {
+                    if ((opn2->ch_out[ii]!=old_out_fm[ii])) {
+                        int freq=opn2->fnum[ii];
+                        int octave=opn2->block[ii];
+                        vgm_last_note[ii+m_voice_ofs]=(freq<<octave)*110.0f/1081.0f; //1148.0f;
+                        vgm_last_instr[ii+m_voice_ofs]=ii+m_voice_ofs;
+                        int newvol=1;//cch[ii]->keyon_triggered+1;
+                            vgm_last_vol[ii+m_voice_ofs]=newvol;
+                    }
+                }
+            }
+        }
+    
+    for (int ii=0;ii<6;ii++) {
+        old_out_fm[ii]=opn2->ch_out[ii];
+        old_dacen=opn2->dacen;
+    }
+    //YOYOFR
 
     for (i = 0; i < numsamples; i++)
     {
