diff --git a/emu/cores/es5503.c b/emu/cores/es5503.c
index a0a2a99..b4b6d70 100644
--- a/emu/cores/es5503.c
+++ b/emu/cores/es5503.c
@@ -50,6 +50,11 @@
 #include "../EmuHelper.h"
 #include "es5503.h"
 
+//TODO:  MODIZER changes start / YOYOFR
+#include "../../../../../src/ModizerVoicesData.h"
+//TODO:  MODIZER changes end / YOYOFR
+
+
 
 static void es5503_w(void *info, UINT8 offset, UINT8 data);
 static UINT8 es5503_r(void *info, UINT8 offset);
@@ -265,6 +270,25 @@ static void es5503_pcm_update(void *param, UINT32 samples, DEV_SMPL **outputs)
 	UINT32 ramptr;
 	ES5503Chip *chip = (ES5503Chip *)param;
 	UINT8 chnsStereo, chan;
+    
+    //TODO:  MODIZER changes start / YOYOFR
+    //search first voice linked to current chip
+    int64_t m_voice_ofs=-1;
+    int m_total_channels=32;
+    for (int ii=0;ii<=SOUND_MAXVOICES_BUFFER_FX-m_total_channels;ii++) {
+        if (m_voice_ChipID[ii]==m_voice_current_system) {
+            m_voice_ofs=ii+(m_voice_current_systemSub?m_voice_current_systemPairedOfs:0);
+            m_voice_current_total=m_total_channels;
+            break;
+        }
+    }
+    m_voice_current_samplerate=es5503_get_sample_rate(chip);
+    if (!m_voice_current_samplerate) {
+        m_voice_current_samplerate=44100;
+        //printf("voice sample rate null\n");
+    }
+    int smplIncr=(int64_t)44100*(1<<MODIZER_OSCILLO_OFFSET_FIXEDPOINT)/m_voice_current_samplerate;
+    //TODO:  MODIZER changes end / YOYOFR
 
 	memset(outputs[0], 0, samples * sizeof(DEV_SMPL));
 	memset(outputs[1], 0, samples * sizeof(DEV_SMPL));
@@ -346,6 +370,24 @@ static void es5503_pcm_update(void *param, UINT32 samples, DEV_SMPL **outputs)
 							outputs[1][snum] += outData;
 						}
 					}
+                    
+                    //TODO:  MODIZER changes start / YOYOFR
+                    if (m_voice_ofs>=0) {
+                        int j=osc;
+                        int64_t ofs_start=m_voice_current_ptr[m_voice_ofs+j];
+                        int64_t ofs_end=(m_voice_current_ptr[m_voice_ofs+j]+smplIncr);
+                        if ((ofs_end)>(ofs_start)) {
+                            int val=outData>>7;
+                            if (val)
+                                for (;;) {
+                                    m_voice_buff_accumul_temp[m_voice_ofs+j][(ofs_start>>MODIZER_OSCILLO_OFFSET_FIXEDPOINT)&(SOUND_BUFFER_SIZE_SAMPLE*2-1)]+=val;
+                                    m_voice_buff_accumul_temp_cnt[m_voice_ofs+j][(ofs_start>>MODIZER_OSCILLO_OFFSET_FIXEDPOINT)&(SOUND_BUFFER_SIZE_SAMPLE*2-1)]++;
+                                    ofs_start+=(1<<MODIZER_OSCILLO_OFFSET_FIXEDPOINT);
+                                    if (ofs_start>=ofs_end) break;
+                                }
+                        }
+                    }
+                    //TODO:  MODIZER changes end / YOYOFR
 
 					if (altram >= wtsize)
 					{
@@ -354,7 +396,33 @@ static void es5503_pcm_update(void *param, UINT32 samples, DEV_SMPL **outputs)
 				}
 			}	// end if (oscillators[osc] playing)
 		}	// end for (osc)
+        
+        
+        //TODO:  MODIZER changes start / YOYOFR
+        if (m_voice_ofs>=0) {
+            for (int j=0;j<32;j++) {
+                int64_t ofs_end=(m_voice_current_ptr[m_voice_ofs+j]+smplIncr);
+                while ((ofs_end>>MODIZER_OSCILLO_OFFSET_FIXEDPOINT)>=SOUND_BUFFER_SIZE_SAMPLE*2) ofs_end-=(SOUND_BUFFER_SIZE_SAMPLE*2<<MODIZER_OSCILLO_OFFSET_FIXEDPOINT);
+                m_voice_current_ptr[m_voice_ofs+j]=ofs_end;
+            }
+        }
+        //TODO:  MODIZER changes end / YOYOFR
 	}	// end for (snum)
+    
+    //TODO:  MODIZER changes start / YOYOFR
+    if (m_voice_ofs>=0) {
+        for (int v=0;v<chip->oscsenabled;v++) {
+            ES5503Osc *pOsc = &chip->oscillators[v];
+            int j=v;
+            if (!pOsc->Muted && pOsc->vol && (pOsc->freq>0)) {
+                double freq=(double)(pOsc->freq)*(double)m_voice_current_samplerate/(1<<(17+pOsc->wavetblsize));
+                vgm_last_note[j+m_voice_ofs]=freq;
+                vgm_last_vol[j+m_voice_ofs]=1;
+                vgm_last_instr[j+m_voice_ofs]=j+m_voice_ofs;
+            }
+        }
+    }
+    //TODO:  MODIZER changes end / YOYOFR
 }
 
 
