diff --git a/emu/cores/pokey.c b/emu/cores/pokey.c
index b6cd749..fc97cfb 100644
--- a/emu/cores/pokey.c
+++ b/emu/cores/pokey.c
@@ -96,6 +96,11 @@
 #include "../EmuHelper.h"
 #include "pokey.h"
 
+//TODO:  MODIZER changes start / YOYOFR
+#include "../../../../../src/ModizerVoicesData.h"
+//TODO:  MODIZER changes end / YOYOFR
+
+
 typedef struct pokey_device pokey_device;
 typedef struct pokey_channel pokey_channel;
 
@@ -982,17 +987,114 @@ void pokey_update(void *info, UINT32 samples, DEV_SMPL **outputs)
 {
 	pokey_device *d = (pokey_device *)info;
 	UINT32 sampindex;
+    
+    //TODO:  MODIZER changes start / YOYOFR
+    //search first voice linked to current chip
+    int m_voice_ofs=-1;
+    int m_total_channels=4;
+    for (int ii=0;ii<=SOUND_MAXVOICES_BUFFER_FX-m_total_channels;ii++) {
+        if (m_voice_ChipID[ii]==m_voice_current_system) {
+            m_voice_ofs=ii+(m_voice_current_systemSub?m_voice_current_systemPairedOfs:0);
+            m_voice_current_total=m_total_channels;
+            break;
+        }
+    }
+    if (!m_voice_current_samplerate) {
+        m_voice_current_samplerate=44100;
+        //printf("voice sample rate null\n");
+    }
+    int64_t smplIncr;
+    int period[4];
+    smplIncr=(int64_t)44100*(1<<MODIZER_OSCILLO_OFFSET_FIXEDPOINT)/m_voice_current_samplerate;
+    //TODO:  MODIZER changes end / YOYOFR
 
 	for(sampindex = 0; sampindex < samples; sampindex++)
 	{
 		pokey_execute_run(d);
+        
+        //YOYOFR
+        if (sampindex==0) {
+            int div=(d->m_AUDCTL&1?114:28);
+            switch (d->m_AUDCTL & 80) {
+                case 0:
+                    period[0] = div* (d->m_channel[0].m_AUDF + 1);
+                    period[1] = div* (d->m_channel[1].m_AUDF + 1);
+                    break;
+                case 16:
+                    period[0] = div << 8;
+                    period[1] = div * (d->m_channel[0].m_AUDF + (d->m_channel[1].m_AUDF << 8) + 1);
+                    break;
+                case 64:
+                    period[0] = d->m_channel[0].m_AUDF + 4;
+                    period[1] = div * (d->m_channel[1].m_AUDF + 1);
+                    break;
+                case 80:
+                    period[0] = 256;
+                    period[1] = d->m_channel[0].m_AUDF + (d->m_channel[1].m_AUDF << 8) + 7;
+                    break;
+            }
+            switch (d->m_AUDCTL & 40) {
+                case 0:
+                    period[2] = div * (d->m_channel[2].m_AUDF + 1);
+                    period[3] = div * (d->m_channel[3].m_AUDF + 1);
+                    break;
+                case 8:
+                    period[2] = div << 8;
+                    period[3] = div * (d->m_channel[2].m_AUDF + (d->m_channel[3].m_AUDF << 8) + 1);
+                    break;
+                case 32:
+                    period[2] = d->m_channel[2].m_AUDF + 4;
+                    period[3] = div * (d->m_channel[3].m_AUDF + 1);
+                    break;
+                case 40:
+                    period[2] = 256;
+                    period[3] = d->m_channel[2].m_AUDF + (d->m_channel[3].m_AUDF << 8) + 7;
+                    break;
+            }
+        }
+        //YOYOFR
 
 		if (d->m_output_type == LEGACY_LINEAR)
 		{
 			INT32 out = 0;
 			int i;
-			for (i = 0; i < 4; i++)
-				out += ((d->m_out_raw >> (4*i)) & 0x0f);
+            for (i = 0; i < 4; i++) {
+                out += ((d->m_out_raw >> (4*i)) & 0x0f);
+                
+                //TODO:  MODIZER changes start / YOYOFR
+                if (m_voice_ofs>=0) {
+                    int64_t ofs_start=m_voice_current_ptr[m_voice_ofs+i];
+                    int64_t ofs_end=(m_voice_current_ptr[m_voice_ofs+i]+smplIncr);
+                    if (ofs_end>ofs_start)
+                        for (;;) {
+                            int val=((d->m_out_raw >> (4*i)) & 0x0f);
+                            val *= POKEY_DEFAULT_GAIN;
+                            val = (val > 0x7fff) ? 0x7fff : val;
+                            m_voice_buff[m_voice_ofs+i][(ofs_start>>MODIZER_OSCILLO_OFFSET_FIXEDPOINT)&(SOUND_BUFFER_SIZE_SAMPLE*4*2-1)]=LIMIT8( val>>5 );
+                            ofs_start+=1<<MODIZER_OSCILLO_OFFSET_FIXEDPOINT;
+                            if (ofs_start>=ofs_end) break;
+                        }
+                    m_voice_current_ptr[m_voice_ofs+i]=ofs_end;
+                }
+                //TODO:  MODIZER changes end / YOYOFR
+                //YOYOFR
+                if (sampindex==0) {
+                    if (!(d->m_muted[i])) {
+                        
+                        double freq=period[i];
+                        int vol=((d->m_channel[i].m_AUDC & 15) >0);
+                        if (freq && vol) {
+                            freq=1/d->m_clock_period/freq/2;
+                            vgm_last_note[i+m_voice_ofs]=freq; ;//440.0f*c->v[i].freq/22050.0f;
+                            vgm_last_instr[i+m_voice_ofs]=i+m_voice_ofs;
+                            vgm_last_vol[i+m_voice_ofs]=1;
+                        }
+                    }
+                }
+                //YOYOFR
+            }
+            
+            
 			out *= POKEY_DEFAULT_GAIN;
 			out = (out > 0x7fff) ? 0x7fff : out;
 			outputs[0][sampindex] = out;
