diff --git a/player/gymplayer.cpp b/player/gymplayer.cpp
index 2ee717c..217b576 100644
--- a/player/gymplayer.cpp
+++ b/player/gymplayer.cpp
@@ -18,6 +18,11 @@
 #include "helper.h"
 #include "../emu/logging.h"
 
+//TODO:  MODIZER changes start / YOYOFR
+#include "../../../../src/ModizerVoicesData.h"
+//TODO:  MODIZER changes end / YOYOFR
+
+
 
 INLINE UINT32 ReadLE32(const UINT8* data)
 {
@@ -940,11 +945,25 @@ UINT32 GYMPlayer::Render(UINT32 smplCnt, WAVE_32BS* data)
 			GYM_CHIPDEV* cDev = &_devices[curDev];
 			UINT8 disable = (cDev->optID != (size_t)-1) ? _devOpts[cDev->optID].muteOpts.disable : 0x00;
 			VGM_BASEDEV* clDev;
+            //TODO:  MODIZER changes start / yoyofr
+            m_voice_current_system=curDev;
+            m_voice_current_systemSub=0;  //YOYOFR: to remove, not used anymore
+            m_voice_current_systemPairedOfs=0;
+            //TODO:  MODIZER changes end / YOYOFR
 			
 			for (clDev = &cDev->base; clDev != NULL; clDev = clDev->linkDev, disable >>= 1)
 			{
+                //YOYOFR
+                m_voice_current_samplerate=clDev->defInf.sampleRate;
+                //YOYOFR
+                
 				if (clDev->defInf.dataPtr != NULL && ! (disable & 0x01))
 					Resmpl_Execute(&clDev->resmpl, smplStep, &data[curSmpl]);
+                
+                //YOYOFR
+                m_voice_current_systemSub++; //flag that next one will be a linked device
+                m_voice_current_systemPairedOfs+=m_voice_current_total;
+                //YOYOFR
 			}
 		}
 		curSmpl += smplStep;
