diff --git a/player/droplayer.cpp b/player/droplayer.cpp
index 2bf9312..c884f63 100644
--- a/player/droplayer.cpp
+++ b/player/droplayer.cpp
@@ -16,6 +16,11 @@
 #include "helper.h"
 #include "../emu/logging.h"
 
+//TODO:  MODIZER changes start / YOYOFR
+#include "../../../../src/ModizerVoicesData.h"
+//TODO:  MODIZER changes end / YOYOFR
+
+
 #ifdef _MSC_VER
 #define snprintf	_snprintf
 #endif
@@ -933,11 +938,25 @@ UINT32 DROPlayer::Render(UINT32 smplCnt, WAVE_32BS* data)
 			DRO_CHIPDEV* cDev = &_devices[curDev];
 			UINT8 disable = (cDev->optID != (size_t)-1) ? _devOpts[cDev->optID].muteOpts.disable : 0x00;
 			VGM_BASEDEV* clDev;
+            //TODO:  MODIZER changes start / yoyofr
+            m_voice_current_system=curDev;
+            m_voice_current_systemSub=0;  //YOYOFR: to remove, not used anymore
+            m_voice_current_systemPairedOfs=0;
+            //TODO:  MODIZER changes end / YOYOFR
 			
 			for (clDev = &cDev->base; clDev != NULL; clDev = clDev->linkDev, disable >>= 1)
 			{
+                //YOYOFR
+                m_voice_current_samplerate=clDev->defInf.sampleRate;
+                //YOYOFR
+                
 				if (clDev->defInf.dataPtr != NULL && ! (disable & 0x01))
 					Resmpl_Execute(&clDev->resmpl, smplStep, &data[curSmpl]);
+                
+                //YOYOFR
+                m_voice_current_systemSub++; //flag that next one will be a linked device
+                m_voice_current_systemPairedOfs+=m_voice_current_total;
+                //YOYOFR
 			}
 		}
 		curSmpl += smplStep;
