--- third_party/adplug/upstream/woodyopl.cpp	2026-07-05 17:21:16
+++ third_party/adplug/src/woodyopl.cpp	2026-07-05 13:55:57
@@ -27,6 +27,12 @@
 /*
  * Modified for AdPlug
  */
+
+//TODO:  MODIZER changes start / YOYOFR
+#include "../../../src/ModizerVoicesData.h"
+#include <stdio.h>
+//TODO:  MODIZER changes end / YOYOFR
+
 
 #include <math.h>
 #include <stdlib.h> // rand()
@@ -968,6 +974,11 @@
 	outbufl[i] += chanval;
 #endif
 
+    //YOYOFR
+#define YOYOFR_CHANNELDATA_STORE(channel,voice_op,mulfact) if (m_voice_current_system) {voicesData[channel][i]+=op[voice_op].cval*(mulfact+((adlibreg[0x105]&1)?(cptr[0].left_pan+cptr[0].right_pan):0));}
+#define YOYOFR_MUTE_VOICE(a) if (generic_mute_mask&(1<<((a)%18))) chanval=0;
+    //YOYOFR
+
 void OPLChipClass::adlib_getsample(Bit16s* sndptr, Bits numsamples) {
 	Bits i, endsamples;
 	op_type* cptr;
@@ -984,10 +995,59 @@
 	Bit32s trem_lut[BLOCKBUF_SIZE];
 
 	Bits samples_to_process = numsamples;
+    
+    //TODO:  MODIZER changes start / YOYOFR
+    Bit32s voicesData[NUM_CHANNELS][BLOCKBUF_SIZE];
+    int m_total_channels=18;
+    int64_t smplIncr=1<<MODIZER_OSCILLO_OFFSET_FIXEDPOINT;
+    //TODO:  MODIZER changes end / YOYOFR
+    
+    //YOYOFR
+    memset(vgm_last_note,0,sizeof(vgm_last_note));
+    memset(vgm_last_vol,0,sizeof(vgm_last_vol));
+    //static Bit32s old_out_fm[18];
+        for (int ii=0;ii<18;ii++) {
+            if ( !(generic_mute_mask&(1<<ii))  ) {
+                // frequency
+                int idx=(ii%9)+(ii/9)*0x100;
+                int64_t frn = (((Bit32u)(adlibreg[ARC_KON_BNUM+idx])&3)<<8) + adlibreg[ARC_FREQ_NUM+idx];
+                // block number/octave
+                int64_t oct = ((((Bit32u)adlibreg[ARC_KON_BNUM+idx])>>2)&7);
+                int keyon=((((Bit32u)adlibreg[ARC_KON_BNUM+idx])>>5)&1);
+                
+                if (keyon) {
+                    if (!(frn)) {
+                        //if ((OPL->chan_out[ii] !=old_out_fm[ii]))
+                        {
+                            vgm_last_note[ii]=220.0f; //arbitrary choosing A-3
+                            vgm_last_instr[ii]=ii;
+                            int newvol=keyon;//chip->P_CH[ii].keyon_triggered;//+1;
+                            vgm_last_vol[ii]=newvol;
+                        }
+                    } else {
+                        //if ((OPL->chan_out[ii]!=old_out_fm[ii]))
+                        {
+                            int64_t note=frn*(1<<oct)*INTFREQU/(1<<18);
+                            
+                            vgm_last_note[ii]=note;
+                            vgm_last_instr[ii]=ii;
+                            int newvol=keyon;//chip->P_CH[ii].keyon_triggered;//+1;
+                            vgm_last_vol[ii]=newvol;
+                        }
+                    }
+                }
+                //old_out_fm[ii]=OPL->chan_out[ii];
+            }
+        }
+    //YOYOFR
 
 	for (Bits cursmp=0; cursmp<samples_to_process; cursmp+=endsamples) {
 		endsamples = samples_to_process-cursmp;
 		if (endsamples>BLOCKBUF_SIZE) endsamples = BLOCKBUF_SIZE;
+        
+        //YOYOFR
+        if (m_voice_current_system) {for (int uu=0;uu<NUM_CHANNELS;uu++) memset(voicesData[uu],0,BLOCKBUF_SIZE*sizeof(Bit32s));}
+        //YOYOFR
 
 		memset((void*)&outbufl,0,endsamples*sizeof(Bit32s));
 #if defined(OPLTYPE_IS_OPL3)
@@ -1031,7 +1091,13 @@
 						operator_output(&cptr[9],0,tremval1[i]);
 						
 						Bit32s chanval = cptr[9].cval*2;
+                        
+                        YOYOFR_MUTE_VOICE(6)
 						CHANVAL_OUT
+                        
+                        //TODO:  MODIZER changes start / YOYOFR
+                        YOYOFR_CHANNELDATA_STORE(6,6+9,2)
+                        //TODO:  MODIZER changes end / YOYOFR
 					}
 				}
 			} else {
@@ -1063,7 +1129,13 @@
 						operator_output(&cptr[9],cptr[0].cval*FIXEDPT,tremval2[i]);
 						
 						Bit32s chanval = cptr[9].cval*2;
+                        
+                        YOYOFR_MUTE_VOICE(6)
 						CHANVAL_OUT
+                        
+                        //TODO:  MODIZER changes start / YOYOFR
+                        YOYOFR_CHANNELDATA_STORE(6,6+9,2)
+                        //TODO:  MODIZER changes end / YOYOFR
 					}
 				}
 			}
@@ -1086,7 +1158,13 @@
 					opfuncs[cptr[0].op_state](&cptr[0]);		//TomTom
 					operator_output(&cptr[0],0,tremval3[i]);
 					Bit32s chanval = cptr[0].cval*2;
+                    
+                    YOYOFR_MUTE_VOICE(8)
 					CHANVAL_OUT
+                    
+                    //TODO:  MODIZER changes start / YOYOFR
+                    YOYOFR_CHANNELDATA_STORE(8,8+0,2)
+                    //TODO:  MODIZER changes end / YOYOFR
 				}
 			}
 
@@ -1134,7 +1212,17 @@
 					operator_output(&op[8+9],0,tremval4[i]);
 
 					Bit32s chanval = (op[7].cval + op[7+9].cval + op[8+9].cval)*2;
+                    
+                    YOYOFR_MUTE_VOICE(7)
+                    //YOYOFR_MUTE_VOICE(7+9)
+                    YOYOFR_MUTE_VOICE(8)
 					CHANVAL_OUT
+                    
+                    //TODO:  MODIZER changes start / YOYOFR
+                    YOYOFR_CHANNELDATA_STORE(7,7,2)
+                    YOYOFR_CHANNELDATA_STORE(7,7+9,2)
+                    YOYOFR_CHANNELDATA_STORE(8,8+9,2)
+                    //TODO:  MODIZER changes end / YOYOFR
 				}
 			}
 		}
@@ -1146,14 +1234,24 @@
 		for (Bits cur_ch=max_channel-1; cur_ch>=0; cur_ch--) {
 			// skip drum/percussion operators
 			if ((adlibreg[ARC_PERC_MODE]&0x20) && (cur_ch >= 6) && (cur_ch < 9)) continue;
+            
+            //TODO:  MODIZER changes start / YOYOFR
+            int channel_op;
+            //TODO:  MODIZER changes end / YOYOFR
 
 			Bitu k = cur_ch;
 #if defined(OPLTYPE_IS_OPL3)
 			if (cur_ch < 9) {
 				cptr = &op[cur_ch];
+                //TODO:  MODIZER changes start / YOYOFR
+                channel_op=cur_ch;
+                //TODO:  MODIZER changes end / YOYOFR
 			} else {
 				cptr = &op[cur_ch+9];	// second set is operator18-operator35
 				k += (-9+256);		// second set uses registers 0x100 onwards
+                //TODO:  MODIZER changes start / YOYOFR
+                channel_op=cur_ch+9;
+                //TODO:  MODIZER changes end / YOYOFR
 			}
 			// check if this operator is part of a 4-op
 			if ((adlibreg[0x105]&1) && cptr->is_4op_attached) continue;
@@ -1183,7 +1281,13 @@
 								operator_output(&cptr[0],(cptr[0].lastcval+cptr[0].cval)*cptr[0].mfbi/2,tremval1[i]);
 
 								Bit32s chanval = cptr[0].cval;
-								CHANVAL_OUT
+								
+                                YOYOFR_MUTE_VOICE(cur_ch)
+                                CHANVAL_OUT
+                                
+                                //YOYOFR
+                                YOYOFR_CHANNELDATA_STORE(cur_ch,channel_op+0,1)
+                                //YOYOFR
 							}
 						}
 
@@ -1209,7 +1313,13 @@
 								operator_output(&cptr[3],cptr[9].cval*FIXEDPT,tremval2[i]);
 
 								Bit32s chanval = cptr[3].cval;
-								CHANVAL_OUT
+								
+                                YOYOFR_MUTE_VOICE(cur_ch)
+                                CHANVAL_OUT
+                                
+                                //YOYOFR
+                                YOYOFR_CHANNELDATA_STORE(cur_ch,channel_op+3,1)
+                                //YOYOFR
 							}
 						}
 
@@ -1224,7 +1334,13 @@
 								operator_output(&cptr[3+9],0,tremval1[i]);
 
 								Bit32s chanval = cptr[3+9].cval;
-								CHANVAL_OUT
+								
+                                YOYOFR_MUTE_VOICE(cur_ch)
+                                CHANVAL_OUT
+                                
+                                //YOYOFR
+                                YOYOFR_CHANNELDATA_STORE(cur_ch,channel_op+3+9,1)
+                                //YOYOFR
 							}
 						}
 					} else {
@@ -1245,7 +1361,13 @@
 								operator_output(&cptr[0],(cptr[0].lastcval+cptr[0].cval)*cptr[0].mfbi/2,tremval1[i]);
 
 								Bit32s chanval = cptr[0].cval;
-								CHANVAL_OUT
+								
+                                YOYOFR_MUTE_VOICE(cur_ch)
+                                CHANVAL_OUT
+                                
+                                //YOYOFR
+                                YOYOFR_CHANNELDATA_STORE(cur_ch,channel_op+0,1)
+                                //YOYOFR
 							}
 						}
 
@@ -1277,7 +1399,13 @@
 								operator_output(&cptr[3+9],cptr[3].cval*FIXEDPT,tremval3[i]);
 
 								Bit32s chanval = cptr[3+9].cval;
-								CHANVAL_OUT
+								
+                                YOYOFR_MUTE_VOICE(cur_ch)
+                                CHANVAL_OUT
+                                
+                                //YOYOFR
+                                YOYOFR_CHANNELDATA_STORE(cur_ch,channel_op+3+9,1)
+                                //YOYOFR
 							}
 						}
 					}
@@ -1314,7 +1442,15 @@
 					operator_output(&cptr[9],0,tremval2[i]);
 
 					Bit32s chanval = cptr[9].cval + cptr[0].cval;
-					CHANVAL_OUT
+					
+                    YOYOFR_MUTE_VOICE(cur_ch)
+                    //YOYOFR_MUTE_VOICE(voice_channel+9)
+                    CHANVAL_OUT
+                    
+                    //YOYOFR
+                    YOYOFR_CHANNELDATA_STORE(cur_ch,channel_op+0,1)
+                    YOYOFR_CHANNELDATA_STORE(cur_ch,channel_op+9,1)
+                    //YOYOFR
 				}
 			} else {
 #if defined(OPLTYPE_IS_OPL3)
@@ -1348,7 +1484,13 @@
 								operator_output(&cptr[9],cptr[0].cval*FIXEDPT,tremval2[i]);
 
 								Bit32s chanval = cptr[9].cval;
-								CHANVAL_OUT
+								
+                                YOYOFR_MUTE_VOICE(cur_ch)
+                                CHANVAL_OUT
+                                
+                                //YOYOFR
+                                YOYOFR_CHANNELDATA_STORE(cur_ch,channel_op+9,1)
+                                //YOYOFR
 							}
 						}
 
@@ -1369,7 +1511,13 @@
 								operator_output(&cptr[3+9],cptr[3].cval*FIXEDPT,tremval2[i]);
 
 								Bit32s chanval = cptr[3+9].cval;
-								CHANVAL_OUT
+								
+                                YOYOFR_MUTE_VOICE(cur_ch)
+                                CHANVAL_OUT
+                                
+                                //YOYOFR
+                                YOYOFR_CHANNELDATA_STORE(cur_ch,channel_op+3+9,1)
+                                //YOYOFR
 							}
 						}
 
@@ -1415,7 +1563,13 @@
 								operator_output(&cptr[3+9],cptr[3].cval*FIXEDPT,tremval4[i]);
 
 								Bit32s chanval = cptr[3+9].cval;
-								CHANVAL_OUT
+								
+                                YOYOFR_MUTE_VOICE(cur_ch)
+                                CHANVAL_OUT
+                                
+                                //YOYOFR
+                                YOYOFR_CHANNELDATA_STORE(cur_ch,channel_op+3+9,1)
+                                //YOYOFR
 							}
 						}
 					}
@@ -1452,7 +1606,13 @@
 					operator_output(&cptr[9],cptr[0].cval*FIXEDPT,tremval2[i]);
 
 					Bit32s chanval = cptr[9].cval;
-					CHANVAL_OUT
+					
+                    YOYOFR_MUTE_VOICE(cur_ch)
+                    CHANVAL_OUT
+                    
+                    //YOYOFR
+                    YOYOFR_CHANNELDATA_STORE(cur_ch,channel_op+9,1)
+                    //YOYOFR
 				}
 			}
 		}
@@ -1507,6 +1667,25 @@
 				}
 			}
 		}
+        
+        //TODO:  MODIZER changes start / YOYOFR
+        if (m_voice_current_system) {
+            for (int uu=0;uu<NUM_CHANNELS;uu++) {
+                for (int i=0;i<endsamples;i++) {
+                    int64_t ofs_start=m_voice_current_ptr[uu];
+                    int64_t ofs_end=(m_voice_current_ptr[uu]+smplIncr);
+                    for (;;) {
+                        if (generic_mute_mask&(1<<uu)) m_voice_buff[uu][(ofs_start>>MODIZER_OSCILLO_OFFSET_FIXEDPOINT)&(SOUND_BUFFER_SIZE_SAMPLE*2-1)]=0;
+                        else m_voice_buff[uu][(ofs_start>>MODIZER_OSCILLO_OFFSET_FIXEDPOINT)&(SOUND_BUFFER_SIZE_SAMPLE*2-1)]=LIMIT8(voicesData[uu][i]>>7);
+                        ofs_start+=1<<MODIZER_OSCILLO_OFFSET_FIXEDPOINT;
+                        if (ofs_start>=ofs_end) break;
+                    }
+                    while ((ofs_end>>MODIZER_OSCILLO_OFFSET_FIXEDPOINT)>=SOUND_BUFFER_SIZE_SAMPLE*2) ofs_end-=(SOUND_BUFFER_SIZE_SAMPLE*2<<MODIZER_OSCILLO_OFFSET_FIXEDPOINT);
+                    m_voice_current_ptr[uu]=ofs_end;
+                }
+            }
+        }
+        //TODO:  MODIZER changes end / YOYOFR
 
 	}
 }
